sklearn/neighbors/quad_tree.pyx

```diff
@@ -0,0 +1,610 @@
+            else:
+                child.min_bounds[i] = cell.min_bounds[i]
+                child.max_bounds[i] = cell.center[i]
+            child.center[i] = (child.min_bounds[i] + child.max_bounds[i]) / 2.
+            width = child.max_bounds[i] - child.min_bounds[i]
+
+            child.barycenter[i] = point[i]
+            child.max_width = max(child.max_width, width*width)
+
+        # Store the point info and the size to account for duplicated points
+        child.point_index = point_index
+        child.cumulative_size = size
+
+        # Store the child cell in the correct place in children
+        cell.children[cell_child_id] = child.cell_id
+
+        if DEBUGFLAG:
+            # Assert that the point is in the right range
+            self.check_point_in_cell(point, child)
+        if self.verbose >= 10:
+            printf("[QuadTree] inserted point %li in new child %li\n",
+                   point_index, cell_id)
+
+        return cell_id
+
+
+    cdef bint is_duplicate(self, DTYPE_t[3] point1, DTYPE_t[3] point2) nogil:
+        """Check if the two given points are equals."""
+        cdef int i
+        cdef bint res = True
+        for i in range(self.n_dimensions):
+            res &= fabsf(point1[i] - point2[i]) <= EPSILON
+        return res
+
+
+    cdef SIZE_t select_child(self, DTYPE_t[3] point, Cell* cell) nogil:
+        """Select the child of cell which contains the given query point."""
+        cdef:
+            int i
+            SIZE_t selected_child = 0
+
+        for i in range(self.n_dimensions):
+            # Select the correct child cell to insert the point by comparing
+            # it to the borders of the cells using precomputed center.
+            selected_child *= 2
+            if point[i] >= cell.center[i]:
+                selected_child += 1
+        return cell.children[selected_child]
+
+    cdef void _init_cell(self, Cell* cell, SIZE_t parent, SIZE_t depth) nogil:
+        """Initialize a cell structure with some constants."""
+        cell.parent = parent
+        cell.is_leaf = True
+        cell.depth = depth
+        cell.max_width = 0
+        cell.cumulative_size = 0
+        for i in range(self.n_cells_per_cell):
+            cell.children[i] = DEFAULT
+
+    cdef void _init_root(self, DTYPE_t[3] min_bounds, DTYPE_t[3] max_bounds
+                         ) nogil:
+        """Initialize the root node with the given space boundaries"""
+        cdef:
+            int i
+            DTYPE_t width
+            Cell* root = &self.cells[0]
+
+        self._init_cell(root, -1, 0)
+        for i in range(self.n_dimensions):
+            root.min_bounds[i] = min_bounds[i]
+            root.max_bounds[i] = max_bounds[i]
+            root.center[i] = (max_bounds[i] + min_bounds[i]) / 2.
+            width = max_bounds[i] - min_bounds[i]
+            root.max_width = max(root.max_width, width*width)
+        root.cell_id = 0
+
+        self.cell_count += 1
+
+    def plot_tree(self):
```

Comment:
On top of the general feeling that this plotting code shouldn't be there, it is also not tested. Testing plotting code is a bit hard. That makes me think that, for the moment, it should be in an example.